### ISSUE TITLE
SyntaxGenerator should not create a TupleTypeSyntax from an unbound generic ITypeSymbol

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/roslyn#67341]
**Version Used**: 4.6.0-2.23081.23

SyntaxGenerator should not create a `TupleTypeSyntax` from an `INamedTypeSymbol` with `IsUnboundGenericType=true`.

Example type symbol: `System.ValueTuple<,,,,,,,>`

**Expected Behavior**:
`(TypeSyntax)_syntaxGenerator.TypeExpression(symbol);` is of type `QualifiedNameSyntax` and doesn't use the tuple type syntax.

**Actual Behavior**:
`(TypeSyntax)_syntaxGenerator.TypeExpression(symbol);` is of type `TupleTypeSyntax` and uses the tuple type syntax.

cc @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
This appears to be an error in the compiler.  Looping in @jcouv .  If we get a symbol from source corresponding to `typeof(ValueTuple<,>)` we get a proper tuple type back with `IsUnboundGenericType` set.  We then properly generate `ValueTuple<,>` on our end.

However, in Viktor's case, the symbol is coming from metadata.  And in metadata that bit seems to be lost.  So we generate a normal `(..., ...)` tuple which doesn't correspond to original source.

This seems to be a bug where the metadata representation doesn't match the source here.  Need compiler to look into this.

**Comment by CyrusNajmabadi** [comment]
Example of original source: https://github.com/dotnet/corefx/blob/d6173e069a9bcedfdfd7f4f41e67d23f67157b61/src/System.ValueTuple/ref/System.ValueTuple.TypeForwards.cs#L5-L15

**Comment by ericstj** [comment]
Probably the bug is due to this hardcoding:
https://github.com/dotnet/roslyn/blob/248e85149427c534c4a156a436ecff69bab83b59/src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs#L1254-L1268

I added a workaround for this bug: https://github.com/dotnet/sdk/pull/32014

**Comment by ericstj** [comment]
Now I'm less certain if this is really a bug or if it's a problem with our usage of the `GetForwardedTypes` API.

What I noticed was that we'd correctly get `IsUnboundGenericType` symbols when examining attributes, it's only when calling `IAssemblySymbol.GetForwardedTypes()` that the resultant types are not unbound generics.  This appears to be due to the implementation of `GetForwardedTypes` which seems to resolve the type definition.

I wrote up a sample app, attached [testSyntaxTypeOf.zip](https://github.com/dotnet/roslyn/files/11327326/testSyntaxTypeOf.zip), that illustrates this:
```c#
using System.Runtime.CompilerServices;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using System.Reflection;

[assembly: TypeForwardedTo(typeof(ValueTuple<,,>))]
[assembly: TypeForwardedTo(typeof(Func<,,>))]
[assembly: TypeForwardedTo(typeof(Action<,,,,>))]
[assembly: Test(typeof(ValueTuple<,,>))]
[assembly: Test(typeof(Func<,,>))]
[assembly: Test(typeof(Action<,,,,>))]

MetadataReference self = MetadataReference.CreateFromFile(Assembly.GetExecutingAssembly().Location);

// make sure project sets `PreserveCompilationContext = true`
MetadataReference corelib = MetadataReference.CreateFromFile(Path.Combine(AppContext.BaseDirectory, "refs", "System.Runtime.dll"));

var cSharpCompilation = CSharpCompilation.Create($"AssemblyLoader_{DateTime.Now:MM_dd_yy_HH_mm_ss_FFF}")
    .AddReferences(self, corelib);
IAssemblySymbol thisAssembly = cSharpCompilation.GetAssemblyOrModuleSymbol(self) as IAssemblySymbol;
var testAttributeArgTypes = thisAssembly.GetAttributes()
    .Where(attr => attr.AttributeClass.Name == nameof(TestAttribute))
    .Select(attr => attr.ConstructorArguments.Single().Value as INamedTypeSymbol);

Console.WriteLine("## Test attributes ##");
PrintTypes(testAttributeArgTypes);  // underlying types are ConstructedNamedTypeSymbol

Console.WriteLine("## Type forwards ##");
PrintTypes(thisAssembly.GetForwardedTypes());  // underlying types are PENamedTypeSymbolGeneric

void PrintTypes(IEnumerable<INamedTypeSymbol> types) { foreach(var type in types) {  Console.WriteLine($"{type} IsUnboundGenericType {type.IsUnboundGenericType}"); } }

[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
public class TestAttribute : Attribute
{
    public TestAttribute(Type type) { }
}
```

The output is:
```
## Test attributes ##
System.ValueTuple<,,> IsUnboundGenericType True
System.Func<,,> IsUnboundGenericType True
System.Action<,,,,> IsUnboundGenericType True
## Type forwards ##
System.Action<T1, T2, T3, T4, T5> IsUnboundGenericType False
System.Func<T1, T2, TResult> IsUnboundGenericType False
(T1, T2, T3) IsUnboundGenericType False
```

I was able to workaround this much more easily, I just added the following extension method and make sure to run it on all types returned from `GetExportedTypes`:
```c#
        public static INamedTypeSymbol MakeUnboundIfGeneric(this INamedTypeSymbol namedType) =>
            namedType.IsGenericType && !namedType.IsUnboundGenericType ? namedType.ConstructUnboundGenericType() : namedType;

```

**Comment by ericstj** [comment]
TL;DR here are some options for roslyn to consider:
1. Ensure that `GetForwardedTypes` returns unbound generic types.
2. Ensure that `SyntaxGenerator.TypeExpression` converts any bound generics to unbound generic types.
3. Make `SyntaxGenerator.TypeExpression` throw if folks pass in a bound generic and make them convert it themselves.
